File: proxy/src/main/java/one/hyro/spark/proxy/commands/TempBanCommand.java
package one.hyro.spark.proxy.commands;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.tree.LiteralCommandNode;
import com.velocitypowered.api.command.BrigadierCommand;
import com.velocitypowered.api.command.CommandSource;
import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

public final class TempBanCommand {

    public static BrigadierCommand createBrigadierCommand(final ProxyServer proxy) {
        LiteralCommandNode<CommandSource> tempBanNode = BrigadierCommand.literalArgumentBuilder("tempban")
                .requires(source -> source.hasPermission("spark.tempban"))
                .executes(context -> {
                    CommandSource source = context.getSource();
                    source.sendMessage(Component.translatable("context.error.specifyPlayer", NamedTextColor.RED));
                    return Command.SINGLE_SUCCESS;
                })
                .then(BrigadierCommand.requiredArgumentBuilder("player", StringArgumentType.word())
                        .suggests((context, builder) -> {
                            List<String> onlinePlayers = proxy.getAllPlayers().stream().map(Player::getUsername).toList();
                            onlinePlayers.forEach(builder::suggest);
                            return builder.buildFuture();
                        })
                        .executes(context -> {
                            CommandSource source = context.getSource();
                            source.sendMessage(Component.translatable("context.error.specifyTime", NamedTextColor.RED));
                            return Command.SINGLE_SUCCESS;
                        })
                        .then(BrigadierCommand.requiredArgumentBuilder("time", StringArgumentType.word())
                                .executes(context -> {
                                    CommandSource source = context.getSource();
                                    source.sendMessage(Component.translatable("context.error.specifyReason", NamedTextColor.RED));
                                    return Command.SINGLE_SUCCESS;
                                })
                                .then(BrigadierCommand.requiredArgumentBuilder("reason", StringArgumentType.word())
                                        .suggests((context, builder) -> {
                                            // Suggest reasons from a list
                                            return builder.buildFuture();
                                        })
                                        .executes(context -> {
                                            CommandSource source = context.getSource();
                                            String player = context.getArgument("player", String.class);
                                            String reason = context.getArgument("reason", String.class);
                                            String time = context.getArgument("time", String.class);
                                            Date expiry = parseTime(time);
                                            Player target = proxy.getPlayer(player).orElse(null);

                                            if (target == null) {
                                                source.sendMessage(Component.translatable("context.error.playerOffline", NamedTextColor.RED));
                                                return Command.SINGLE_SUCCESS;
                                            }

                                            if (expiry == null) {
                                                source.sendMessage(Component.translatable("context.error.invalidTimeFormat", NamedTextColor.RED));
                                                return Command.SINGLE_SUCCESS;
                                            }

                                            Component successMessage = Component.translatable(
                                                    "context.success.ban",
                                                    Component.text(target.getUsername(), NamedTextColor.DARK_GREEN)
                                            ).color(NamedTextColor.GREEN);

                                            // Call API to save ban to database
                                            target.disconnect(Component.translatable("context.success.banMessage", NamedTextColor.RED));
                                            source.sendMessage(successMessage);
                                            return Command.SINGLE_SUCCESS;
                                        })
                                )
                        )
                )
                .build();

        return new BrigadierCommand(tempBanNode);
    }

    private static Date parseTime(String time) {
        String regex = "(\\d+)([dhms])";
        boolean matches = time.matches(regex);
        String unit = time.substring(time.length() - 1);
        int amount = Integer.parseInt(time.substring(0, time.length() - 1));

        if (!matches) return null;
        Calendar calendar = Calendar.getInstance();
        switch (unit) {
            case "d" -> calendar.add(Calendar.DAY_OF_MONTH, amount);
            case "h" -> calendar.add(Calendar.HOUR, amount);
            case "m" -> calendar.add(Calendar.MINUTE, amount);
            case "s" -> calendar.add(Calendar.SECOND, amount);
            default -> {
                return null;
            }
        }

        return calendar.getTime();
    }
}

File: proxy/src/main/kotlin/commands/TempBanCommand.kt
package one.hyro.commands

import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.tree.LiteralCommandNode
import com.velocitypowered.api.command.BrigadierCommand
import com.velocitypowered.api.command.CommandSource
import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import java.util.*

object TempBanCommand {
    private val banReasons = mapOf(
        "Hacking" to "moderation.reasons.hacking",
        "Inappropriate" to "moderation.reasons.inappropriate",
        "Other" to "moderation.reasons.other",
        "Spamming" to "moderation.reasons.spamming",
        "Toxicity" to "moderation.reasons.toxicity"
    )

    fun createTempBanCommand(proxyServer: ProxyServer): BrigadierCommand {
        val tempBanCommandNode: LiteralCommandNode<CommandSource> = BrigadierCommand.literalArgumentBuilder("tempban")
            .requires { source -> source.hasPermission("hyro.tempban") }
            .executes { context ->
                val commandSource: CommandSource = context.source
                commandSource.sendMessage(Component.translatable("context.error.specifyPlayer", NamedTextColor.RED))
                return@executes 0
            }
            .then(BrigadierCommand.requiredArgumentBuilder("player", StringArgumentType.word())
                .suggests { _, builder ->
                    val onlinePlayers = proxyServer.allPlayers.map { it.username }
                    onlinePlayers.forEach { builder.suggest(it) }
                    builder.buildFuture()
                }
                .executes { context ->
                    val commandSource: CommandSource = context.source
                    commandSource.sendMessage(Component.translatable("context.error.specifyTime", NamedTextColor.RED))
                    return@executes 1
                }
                .then(BrigadierCommand.requiredArgumentBuilder("time", StringArgumentType.word())
                    .executes { context ->
                        val commandSource: CommandSource = context.source
                        commandSource.sendMessage(Component.translatable("context.error.specifyReason", NamedTextColor.RED))
                        return@executes 1
                    }
                    .then(BrigadierCommand.requiredArgumentBuilder("reason", StringArgumentType.word())
                        .suggests { _, builder ->
                            banReasons.forEach { (reason, _) -> builder.suggest(reason) }
                            builder.buildFuture()
                        }
                        .executes { context ->
                            val commandSource: CommandSource = context.source
                            val playerName = StringArgumentType.getString(context, "player")
                            val timeArg = StringArgumentType.getString(context, "time")
                            val reason = StringArgumentType.getString(context, "reason")
                            val targetPlayer = proxyServer.getPlayer(playerName)

                            if (reason !in banReasons.keys) {
                                commandSource.sendMessage(Component.translatable("context.errors.invalidReason", NamedTextColor.RED))
                                return@executes 1
                            }

                            if (targetPlayer.isPresent) {
                                val expirationDate = parseDuration(timeArg)
                                if (expirationDate != null) tempBanPlayer(targetPlayer.get(), commandSource, expirationDate, reason)
                                else commandSource.sendMessage(Component.translatable("context.errors.invalidTimeFormat", NamedTextColor.RED))
                            } else {
                                commandSource.sendMessage(Component.translatable("context.error.playerOffline", NamedTextColor.RED))
                            }
                            return@executes 1
                        }
                    )
                )
            )
            .build()

        return BrigadierCommand(tempBanCommandNode)
    }

    private fun parseDuration(timeArg: String): Date? {
        val regex = Regex("(\\d+)([dhms])")
        val matchResult = regex.matchEntire(timeArg) ?: return null
        val (amount, unit) = matchResult.destructured
        val timeAmount = amount.toInt()

        val calendar = Calendar.getInstance()
        when (unit) {
            "d" -> calendar.add(Calendar.DAY_OF_MONTH, timeAmount)
            "h" -> calendar.add(Calendar.HOUR, timeAmount)
            "m" -> calendar.add(Calendar.MINUTE, timeAmount)
            "s" -> calendar.add(Calendar.SECOND, timeAmount)
            else -> return null
        }
        return calendar.time
    }

    private fun tempBanPlayer(player: Player, commandSource: CommandSource, expirationDate: Date, reason: String) {
        // TODO: Add ban to database
        player.disconnect(Component.translatable("context.success.banMessage", NamedTextColor.RED))
        val successMessage = Component.translatable(
            "context.success.ban",
            Component.text(player.username, NamedTextColor.DARK_GREEN)
        ).color(NamedTextColor.GREEN)
        commandSource.sendMessage(successMessage)
    }
}

